Added UDP client example application.

jimenezjose opened the pull request:
# New Features
- Added a C++ UDP client example

jimenezjose commented:
Hey @adamshapiro0 is there a reason I am have 6 tests with status expected?

adamshapiro0 commented:
Yep sorry about that, it was waiting on me. It needs approval to run the tests for first time contributors, but apparently Github doesn't actually send any sort of notification when it is waiting.

FYI, we use a `git rebase` flow to make sure that all commits and merges are 100% linear for testing purposes (you'll see what I mean if you look at the git history with a GUI or with `git log --all --graph --decorate`). Would you mind please rebasing your branch on top of the latest `master` and removing the merge commit that's currently in there?

Once it's rebased let me know and I can look it over.

jimenezjose commented:
@adamshapiro0 this `git rebase` process has not been straight forward. Can I get assistance one squashing all these commit messages onto just one? 

`git checkout udp-cpp-example`

Here is my attempt:
`git log --all --graph --decorate`
I want to combine commits:

* commit 158f776aa7f36da08b1126f798cd5f9b866db9d2 (origin/udp-cpp-example)
    `raw message output typo fix for debug mode`
*  commit 4289a2a2f2d82cae33b35e18005bb9bfe279abf5
    `Merge branch 'PointOneNav:master' into udp-cpp-example`
* commit 1448ec1b88c7fa935d1c17a60e0ff92571b4d34d
    `c++ udp example all setup`

I then did:
`git log --oneline`
To find the commit before `c++ udp example all setup`.

Result: 
1448ec1 `c++ udp example all setup`
5432d73 `Enabled style checking and Python unit tests in CI, and added contributor instructions.`

I then did:
`git rebase -i  5432d73`

such that:
pick e545e9b `Fixed UNKNOWN satellite type value in Python code.`
pick 1448ec1 `c++ udp example all setup`
squash 158f776 `raw message output typo fix for debug mode`

I couldn't find the merge commit but this is what I did to squash 158f776 to 1448ec1.

I then:
`git push --force origin udp-cpp-example`


Should I have removed your commits from being picked? Not sure what to do after this?

adamshapiro0 commented:
Sorry about that, it looks like the rebase may have gotten confused since you already had an unexpected merge commit in the middle of your branch. Our instructions could also use a bit of sprucing up for clarity.

There are really two totally separate operations here:
1. Rebasing your branch to update it with the latest upstream changes
2. Doing an _interactive_ rebase to squash your 2 commits together

The steps you described are correct for doing (2), but I believe the merge commit in between them may have thrown a wrench in the works. Normally that merge wouldn't be there and what you did would work just fine. You can read the command you wrote (`git rebase -i 5432d73`) as "do an interactive rebase and modify all commits starting _after_ commit 5432d73".

In order to update to the latest changes, you want to do (1). For that, you normally want to do `git rebase PointOneNav/master` (I used `PointOneNav/` here since your local repo is cloned from your fork so `origin/` is your fork on github). You'd read that as "rebase all commits on my current branch on top of the current version of `PointOneNav/master`".

Basically:
```
git checkout udp-cpp-example    # Checkout your branch
git fetch PointOneNav           # Fetch the latest changes from the upstream repo
git rebase PointOneNav/master   # Rebase your branch onto the latest version of the upstream repo
```

There are a few easy ways you can correct your branch. Since you already squashed stuff and now just have the one commit (currently 247c428), I'd suggest simply resetting your branch to sit at `PointOneNav/master` and then cherry picking your one commit:

```
git checkout udp-cpp-example
git fetch PointOneNav
git reset --hard PointOneNav/master
git cherry-pick 247c428
git push --force
```

adamshapiro0 commented:
@jimenezjose if you have a few minutes, would you mind taking a look at https://github.com/PointOneNav/fusion-engine-client/pull/72 (rendered version here: https://github.com/PointOneNav/fusion-engine-client/blob/instructions/CONTRIBUTING.md#updating-to-the-latest-changes-rebasing) to see if it's a little more clear? I appreciate any feedback. Thanks!

jimenezjose commented:
@adamshapiro0 
I will documenting my entire process of success:

`git fetch PointOneNav` did not work. 

I resolved this with executing: 
`git remote add PointOneNav https://github.com/PointOneNav/fusion-engine-client.git`
before referencing 'PointOneNav'.

While rebasing to master there was a lot of manual work. Thankfully VSCode helped with this process.

Actually, it's a lot of manual work. Is there a way to automate the rebasing of 91 commits without breaking anything?

adamshapiro0 commented:
@jimenezjose that's helpful, I appreciate your patience and appreciate you documenting it :)

You should definitely not be rebasing 91 commits. You only have 1 commit. As I mentioned, since you already merged into your branch and then rebased, the branch is in a slightly strange state and the rebase may be getting confused. If you notice when it's getting stuck, it's actually now trying to recreate all of _my_ commits old, which already exist on `master`. That's definitely not what you wanted to do.

In this case, I would recommend _not_ rebasing, and instead doing the cherry pick method I suggested:
```
git checkout udp-cpp-example
git fetch PointOneNav
git reset --hard PointOneNav/master
git cherry-pick 247c428              <-- or whatever the correct commit hash is for commit: "c++ udp example all setup"
git push --force
```

Or, another way you could do this if you did want to use rebase is to first reset your branch to erase the merge commit that's currently on top of it, and then use `git rebase --onto` to specifically rebase just your one commit:
```
git checkout udp-cpp-example
git reset --hard 247c428              <-- or whatever the correct commit hash is for commit: "c++ udp example all setup"
git fetch PointOneNav
git rebase --onto PointOneNav/master HEAD~1
```

You can read that as "rebase my branch on top of PointOneNav/master, starting _after_ the commit just before mine (HEAD~1)".

jimenezjose commented:
I will documenting my entire process of success: (continued)

```
git checkout udp-cpp-example
git remote add PointOneNav https://github.com/PointOneNav/fusion-engine-client.git
git fetch PointOneNav
git reset --hard PointOneNav/master
git log --all --graph --decorate                      # used to find/confirm the commit hash for commit: "c++ udp example all setup"
git cherry-pick 247c428
git push --force
```

jimenezjose commented:
@adamshapiro0

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+#include <unistd.h> // For close()
+
+#include <iostream>
+#include <fstream>
```
Sort along with the other includes above please. Standard includes (iostream, fstream) in the first group, system includes (inet.h) in teh second.

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+  }
+}
+
+void * get_in_addr(struct sockaddr * sa) {
```
Formatting of *s not quite right here and below - please run `clang-format` on the file to get all the style stuff.

adamshapiro0 replied:
Add a `/***/` bar before this function, similar to the other functions.

Function names should be title case (`GetInAddr`).

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+    }
+    return &(((struct sockaddr_in6 *)sa)->sin6_addr);
+}
+/******************************************************************************/
```
This bar belongs to `main()`, so please swap this line and the blank line just below.

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+    printf(R"EOF(
+Usage: %s [PORT]
+
+Connect to an Atlas device over UDP and print out the incoming message
```
Unlike TCP, which connects to Atlas as a client, with UDP the data is sent by Atlas to the client. For clarity, I'd suggest rephrasing this more like "Receive incoming data from an Atlas device over UDP and..."

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+
+  int port = (argc == 2) ? std::stoi(argv[1]) : 12345;
+
+  // create UDP socket.
```
Capitalize all comments please :)

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+    return 2;
+  }
+
+  // bind socket to port
```
Same here: capital "Bind" and add a period. We like to make everything complete sentences wherever possible for consistency and readability.

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+
+#include "../common/print_message.h"
+
+#define DEBUG_ON 0
```
This doesn't seem to be used anymore - remove?

jimenezjose replied:
I think this debug statement should stay. Only because as the user of Atlas this helped me debug issues (usually on the server side) about the network topology between the Atlas and the listening linux client. Of course you are the owner of this codebase so it is your decision. But I'd say if you want to make the user's life easier... leave it.

jimenezjose replied:
Also, just thinking out load.. this also shows the developer that it is possible to receive packets from different ip addresses on the same port.

jimenezjose replied:
Oh and also, the compiler would also complain about unused variables (e.g. `their_ip`).

adamshapiro0 replied:
If you found it helpful, I'm ok with leaving it. Just see the notes below about using `#if` instead, and about not printing out `buffer`.

On the unused variables note, technically `recvfrom()` accepts `nullptr` for the `srcaddr` and `addrlen` arguments, so if you are not using them you don't need to specify them, or you can use `recv()` instead. See the note in https://man7.org/linux/man-pages/man2/recv.2.html.

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+
+
+    inet_ntop(their_addr.ss_family, get_in_addr((struct sockaddr *)&their_addr), their_ip, sizeof(their_ip));
+    buffer[bytes_read] = '\0';
```
This incoming data is almost always going to be binary, not a string, so printing it with `%s` below may do unexpected things and probably won't do anything too useful. I imagine this was here from your initial debugging? I'd suggest removing these lines and the if statement/print below now that this is all working. You can also then remove `their_addr` and `their_ip`, and use `recv()` instead of `recvfrom()`.

jimenezjose replied:
As you mentioned before UDP, in contrast to TCP, is connectionless. Usually with TCP you have a connect() call to do the TCP 3-way handshake with the client and server but in this case the server is just spitting out packets with the ip address destined to the client machine. The client machine here just receives the packet on a specified port. That is really the only difference between TCP and UDP. In other words, `recvfrom()` is needed for UDP.

adamshapiro0 replied:
Technically, I believe you can use either `recv()` or `recvfrom()` for UDP if you are not interested in the sender's IP address (`recvfrom()` accepts `nullptr` for the source arguments). From the `recv(2)` man page:

> Also, the following call
>   recv(sockfd, buf, len, flags);
> is equivalent to
>   recvfrom(sockfd, buf, len, flags, NULL, NULL);

In this case, it's probably safe to assume that the packet is from the Atlas device since for UDP you must first configure Atlas to send to your machine specifically, so we don't really need the `their_*` information. That being said, if you think it's useful to keep the debug print then it's probably ok.

In that case though, it's definitely wrap the call in a `#if DEBUG_ON` instead of `if (DEBUG_ON)` -- since `DEBUG_ON` is a constant, using an if statement will produce a `Statement always evaluates to false` warning on some compilers.

jimenezjose replied:
I think that may just work. I just remember that I wanted to do just that in Vegas but for some reason I was having issues using recv() instead of recvfrom(). I would have used recv() but it was not working at the time. And I just remember the TCP client working with those system calls and noticing that the only difference was the connect() instead of bind() for UDP. After reading a little more I think it would be possible to use connect() for UDP transactions and then use recv() and send() bc the source and destination information will automatically be stored. Hmm I'd have to test it though. I wouldn't want to change something that I know works for a code style optimization that has not been tested. Ref: https://beej.us/guide/bgnet/html/index-wide.html#datagram, https://beej.us/guide/bgnet/html/index-wide.html#sendtorecv.

jimenezjose replied:
great point about the #if preprocessor condition.

adamshapiro0 reviewed (changes requested):
Nice work and much appreciated! Just some minor cleanup to take care of and this should be ready for merge.

Feel free to do the cleanup either as a separate commit, or by amending the exiting one if you prefer. Let me know once it's all set and pushed up.

jimenezjose commented:
After resolving the code style I did the following:

```
git checkout udp-cpp-example
git remote add PointOneNav https://github.com/PointOneNav/fusion-engine-client.git
git fetch PointOneNav
git reset --hard PointOneNav/master
git pull
```

*my code style update*

```
git add .
git commit --amend
git push --force origin udp-cpp-example
```

jimenezjose commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,135 @@
+  char their_ip[INET6_ADDRSTRLEN];
+  while (!shutdown_pending_) {
+    ssize_t bytes_read = recvfrom(sock, buffer, sizeof(buffer), 0,
+        (struct sockaddr *)&their_addr, &addr_len);
```
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,136 @@
+    buffer[bytes_read] = '\0';
+    if (DEBUG_ON) {
+      // Enable to debug udp server to client problems.
+      printf("listener: received raw packet [%s] from %s\n", buffer, their_ip);
```
If we are keeping this print, it is very important _not_ to print out `buffer` as a string. It's ok to print `bytes_read` perhaps, but `buffer` contains arbitrary binary data and interpreting it as a string can have unintended consequences.

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,136 @@
+  
+    #if DEBUG_ON
+      // Enable to debug udp server to client problems.
+      printf("listener: received raw packet [%d bytes] from %s\n", bytes_read, their_ip);
```
`bytes_read` is a `size_t` so you should print it with `%zu` instead of `%d`, which is for (signed) `int`.

jimenezjose replied:
done.

adamshapiro0 commented on examples/udp_client/linux_udp_client.cc:
```diff
@@ -0,0 +1,136 @@
+    inet_ntop(their_addr.ss_family, GetInAddr((struct sockaddr*)&their_addr),
+              their_ip, sizeof(their_ip));
+  
+    #if DEBUG_ON
```
You may want rerun `clang-format` on the file. Normally it aligns preprocessor directives at the start of the line like
```
#if DEBUG_ON // yes

     #if DEBUG_ON // no
```

jimenezjose replied:
done.
